src/utilities.typ

```diff
@@ -114,88 +119,124 @@
     previous
 }
 
-// Convert a certain (non-relative) length to pt
+// Convert a length of type length to pt.
 //
 // styles: from style()
-// page_size: equivalent to 100%
-// frac_amount: amount of 'fr' specified
-// frac_total: total space shared by fractions
-#let convert-length-to-pt(
-    len,
-    styles: none, page_size: none, frac_amount: none, frac_total: none
-) = {
-    page_size = 0pt + page_size
+#let convert-length-type-to-pt(len, styles) = {
+    // E.g., "1pt", "1em", "0.5pt", "0.5em", "1pt + 1em", "-0.5pt + -0.5em"
+    // Important notes: there will always be at least one digit before the decimal point
+    // for fractional lengths (e.g., .5em becomes "0.5em"). Negative parts will always
+    // have a single minus sign or hyphen (depending on the Typst version) in front without
+    // any separating space.
+    let len-repr = repr(len)
+
```

Comment:
Re: my comment on using the old algorithm where possible (I believe, without any minus signs, it will always be positive, as it'd be weird to have text with negative size), the code below should be reasonable:

```suggestion
    let len-repr = repr(len)
    if "−" not in len-repr and "-" not in len-repr {
       // No way the length is negative here, so we just measure the line straight away.
       // When the length CAN be negative, measuring the line might not give accurate
       // results, so we perform some extra checks.
       return measure(line(length: len), styles).width
    }
```

Reply:
Actually, I've just considered that we only have to do this regex matching precisely when one of the two components (em / pt) is positive and the other negative. When both are negative, or there is only one and it is negative, you can just multiply the length by -1 and measure, then multiply back by -1. The case where both (or the only one) are positive is the one I propose above.

So perhaps a better solution is to keep track of:
1. If there's an em component (`"em" in len-repr`);
2. If there's a pt component (`"pt" in len-repr`);
3. Perhaps which ones are negative (you could just count the amount of minus characters in the string).

Then, when the amount of minus is 0 (code I suggest above), just measure normally; when the amount of minus matches the amount of components, measure the length multiplied by -1 then multiply the result by -1; and, otherwise, do the regex thing.

That way, we can reduce the need for the "possibly imprecise" approach as much as possible. Tell me what you think about this idea.

Reply:
Sketch:

```suggestion
    let len-repr = repr(len)
    let has-em = "em" in len-repr
    let has-pt = "pt" in len-repr
    if "−" not in len-repr and "-" not in len-repr {
       // No way the length is negative here, so we just measure the line straight away.
       // When the length CAN be negative, measuring the line might not give accurate
       // results, so we perform some extra checks.
       return measure(line(length: len), styles).width
    } else if not has-em or not has-pt or len-repr.matches(regex("−|-")).len() == 2 {
       // All components (em / pt) are negative, so we can just measure the whole length
       // multiplied by -1 and then multiply by -1 again.
       return -measure(line(length: -len), styles).width
    }
    // One component is negative and the other is positive, so we'll need to measure them
    // separately.
```

Reply:
That is a smart idea! This got me thinking: we might be able to avoid regex and the precision problem completely. If I remember correctly, drawing a line with a negative length and measuring it gives a length of 0pt. What we could do is:
- Draw a line with the given length, regardless of whether it's positive or negative.
- If the measured length is more than 0pt, then the given length must have been positive and we simply return the measured length.
- If not (i.e., the measured length is 0pt), the given length must either be 0pt or negative.
- To differentiate, we multiply the given length by -1, then draw and measure a line again.
- If measuring the line still gives 0pt, then the given length must have been 0pt, so we return 0pt.
- Otherwise, it must be negative, so we multiply the measured length by -1 and return it.

Reply:
I did some quick benchmarking using [hyperfine](https://github.com/sharkdp/hyperfine). I'm a rookie at benchmarking, but the results seem okay.

At 464ad1d8d4866909cee03765b6a71864a6d76fd5 (before):
```
Benchmark 1: typst compile tablex-test.typ
  Time (mean ± σ):      4.406 s ±  0.286 s    [User: 4.166 s, System: 0.239 s]
  Range (min … max):    4.101 s …  4.762 s    10 runs
```

At 162f18f3b40a265e19b848a4b7036dfeee167205 (after):
```
Benchmark 1: typst compile tablex-test.typ
  Time (mean ± σ):      4.276 s ±  0.213 s    [User: 4.063 s, System: 0.213 s]
  Range (min … max):    4.085 s …  4.564 s    10 runs
```